spyder_kernels/comms/frontendcomm.py

```diff
@@ -27,8 +43,81 @@ def __init__(self, kernel):
         self.kernel.comm_manager.register_target(
             self._comm_name, self._comm_open)
 
+        self.comm_port = None
+
+        if self.kernel.parent:
+            # self.kernel.parent is IPKernelApp unless we are in tests
+            # Create a new socket
+            context = zmq.Context()
+            self.comm_socket = context.socket(zmq.ROUTER)
+            self.comm_socket.linger = 1000
+
+            self.comm_port = get_port()
+
+            self.comm_port = self.kernel.parent._bind_socket(
+                self.comm_socket, self.comm_port)
+            if hasattr(zmq, 'ROUTER_HANDOVER'):
+                # set router-handover to workaround zeromq reconnect problems
+                # in certain rare circumstances
+                # see ipython/ipykernel#270 and zeromq/libzmq#2892
+                self.comm_socket.router_handover = 1
+
+            self.comm_thread_close = threading.Event()
+            self.comm_socket_thread = threading.Thread(target=self.poll_thread)
+            self.comm_socket_thread.start()
+
+            # patch parent.close
+            if not PY2:
+                super_close = self.kernel.parent.close
+
+                def close():
+                    """Close comm_socket_thread."""
+                    self.comm_thread_close.set()
+                    context.term()
+                    self.comm_socket_thread.join()
+                    super_close()
+
+                self.kernel.parent.close = close
+
+    def poll_thread(self):
+        """Recieve messages from comm socket."""
         if not PY2:
-            self._main_thread_id = threading.get_ident()
+            # Create an event loop to handle some messages
+            ioloop.IOLoop().initialize()
+        while not self.comm_thread_close.is_set():
+            self.poll_one()
+
+    def poll_one(self):
+        """Recieve one message from comm socket."""
+        out_stream = None
+        if self.kernel.shell_streams:
+            out_stream = self.kernel.shell_streams[0]
+        try:
+            ident, msg = self.kernel.session.recv(self.comm_socket, 0)
+        except Exception:
+            self.kernel.log.warning("Invalid Message:", exc_info=True)
+        msg_type = msg['header']['msg_type']
+
+        if msg_type == 'shutdown_request':
+            self.comm_thread_close.set()
+            return
+
+        handler = self.kernel.shell_handlers.get(msg_type, None)
+        if handler is None:
+            self.kernel.log.warning("Unknown message type: %r", msg_type)
+        else:
+            try:
+                # shell_streams[0] handles replies
```

Comment:
Please expand this comment because I don't understand very well why you're handling the comm messages in the `out_stream` (at least that's what it looks like by reading the code below).

Reply:
The `out_stream` is used by the handler to send replies. Usually the shell stream is used to both recieve and send messages. Here, we replace the "recieve" but not the "send", so we still need `out_stream` to send